### DISCUSSION TITLE
how can i emulate sysvipc init?

### DISCUSSION THREAD

**Discussion by dontknowhy** [termux/proot#232] [Q&A]
i notice proot has "--sysvipc" option
but i can launch it and it doesn't works???
when i try to tun "init 3",it allways says:Couldn't find an alternative telinit implementation to spawn.
how can i fix it?
and can i use systemd in Proot?

**Comment by michalbednarski** [comment]
[`sysvipc` stands for System V IPC (Inter Process Communication)](https://man7.org/linux/man-pages/man7/sysvipc.7.html) and is unrelated to `init` (`--sysvipc` switch allows using for example `fakeroot`)

SystemD cannot work inside proot (due to not being pid 1, although I'd expect further issues (haven't tried so I don't know if it'd be cgroups/namespaces/mounts or something else) if that one would be emulated)

Usually proot is used without `init` inside guest and launching necessary services manually from shell

**Reply by dontknowhy** [reply]
but i installed a fully sysvipc init system
and it doesn't works…
why?

**Reply by michalbednarski** [reply]
There is no such thing as "sysvipc init", you mean "SysV init" (without "ipc")

I don't think it really makes sense to do so but you can actually start SysV init using `init -i` command which will make `init` behave as `init` (not `telinit`) despite not being pid 1

**Reply by dontknowhy** [reply]
> There is no such thing as "sysvipc init", you mean "SysV init" (without "ipc")
> 
> I don't think it really makes sense to do so but you can actually start SysV init using `init -i` command which will make `init` behave as `init` (not `telinit`) despite not being pid 1

**_Couldn't find an alternative telinit implementation to spawn._**


did SysV died?
how can i start it normally?
(but i can't use systemd and SysV… dpkg will uninstall my finit…)

**Reply by dontknowhy** [reply]
> > There is no such thing as "sysvipc init", you mean "SysV init" (without "ipc")
> > 
> > I don't think it really makes sense to do so but you can actually start SysV init using `init -i` command which will make `init` behave as `init` (not `telinit`) despite not being pid 1
> 
> **_Couldn't find an alternative telinit implementation to spawn._**
> 
> 
> did SysV died?
> how can i start it normally?
> (but i can't use systemd and SysV… dpkg will uninstall my finit…)

how can i use any system service controler in Proot…

**Reply by michalbednarski** [reply]
"Couldn't find an alternative telinit implementation to spawn." means that you didn't actually install SysV init, only systemd compatibility layer, for `proot-distro debian` installation command would be `apt-mark unhold gvfs-daemons ; apt install sysvinit-core`

Although I'm still recommending not using any `init` at all and manually starting services instead

**Reply by dontknowhy** [reply]
> "Couldn't find an alternative telinit implementation to spawn." means that you didn't actually install SysV init, only systemd compatibility layer, for `proot-distro debian` installation command would be `apt-mark unhold gvfs-daemons ; apt install sysvinit-core`
> 
> Although I'm still recommending not using any `init` at all and manually starting services instead

my termux crashed…
sad story…

**Reply by dontknowhy** [reply]
> > "Couldn't find an alternative telinit implementation to spawn." means that you didn't actually install SysV init, only systemd compatibility layer, for `proot-distro debian` installation command would be `apt-mark unhold gvfs-daemons ; apt install sysvinit-core`
> > 
> > Although I'm still recommending not using any `init` at all and manually starting services instead
> 
> my termux crashed…
> sad story…

my termux started
but initctl was broken(/run/initctl)

**Reply by dontknowhy** [reply]
i can run init now
but how can i start **init -i** and /usr/bin/login at the same time?
(start with proot,exit with proot.just like there isn't any init program i need to start)

**Reply by michalbednarski** [reply]
Start init from shell and continue using shell (wrap in shell script (`init -i & ; login`) if needed)

> like there isn't any init program i need to start

Why do you even want to start init though?

**Reply by dontknowhy** [reply]
> Start init from shell and continue using shell (wrap in shell script (`init -i & ; login`) if needed)
> 
> > like there isn't any init program i need to start
> 
> Why do you even want to start init though?

because it has some service i don't know(?)
and it broken…(not the container,the startup script was broken)
this is the startup script:
#!/data/data/com.termux/files/usr/bin/bash
cd $HOME
## unset LD_PRELOAD in case termux-exec is installed
unset LD_PRELOAD
pulseaudio -k
nice -n 99 ionice -c 2 -n 7 pulseaudio --start --load="module-native-protocol-tcp auth-ip-acl=127.0.0.1 auth-anonymous=1" --exit-idle-time=-1
command="nice -n 99 ionice -c 2 -n 7 proot"
command+=" --link2symlink"
command+=" --kill-on-exit"
command+=" --sysvipc"
command+=" -k 5.17.3-aarch64"
command+=" -0"
command+=" -r debian-fs"
if [ -n "$(ls -A debian-binds)" ]; then
    for f in debian-binds/* ;do
      . $f
    done
fi
command+=" -b /dev"
command+=" -b /proc"
command+=" -b /system"
command+=" -b /storage/emulated/0"
command+=" -b /storage/FE3F-10FC"
command+=" -b /apex"
command+=" -b /vendor"
command+=" -b /odm"
command+=" -b /data"
command+=" -b /sys"
command+=" -b debian-fs/root:/dev/shm"
command+=" -b /data/data/com.termux/files/usr/tmp:/tmp"
command+=" -b /proc/self/fd/2:/dev/stderr"
command+=" -b /proc/self/fd/1:/dev/stdout"
command+=" -b /proc/self/fd/0:/dev/stdin"
command+=" -b /dev/null:/dev/tty0"
command+=" -b /dev/urandom:/dev/random"
command+=" -b /proc/self/fd:/dev/fd"
command+=" -b debian-fs/proc/fakethings/stat:/proc/stat"
command+=" -b debian-fs/proc/fakethings/vmstat:/proc/vmstat"
command+=" -b debian-fs/proc/fakethings/version:/proc/version"
command+=" -b debian-fs/proc/fakethings/bus:/proc/bus"
command+=" -b debian-fs/proc/fakethings/buddyinfo:/proc/buddyinfo"
command+=" -b debian-fs/proc/fakethings/cgroups:/proc/cgroups"
command+=" -b debian-fs/proc/fakethings/consoles:/proc/consoles"
command+=" -b debian-fs/proc/fakethings/crypto:/proc/crypto"
command+=" -b debian-fs/proc/fakethings/devices:/proc/devices"
command+=" -b debian-fs/proc/fakethings/diskstats:/proc/diskstats"
command+=" -b debian-fs/proc/fakethings/execdomains:/proc/execdomains"
command+=" -b debian-fs/proc/fakethings/fb:/proc/fb"
command+=" -b debian-fs/proc/fakethings/filesystems:/proc/filesystems"
command+=" -b debian-fs/proc/fakethings/interrupts:/proc/interrrupts"
command+=" -b debian-fs/proc/fakethings/iomem:/proc/iomem"
command+=" -b debian-fs/proc/fakethings/ioports:/proc/ioports"
command+=" -b debian-fs/proc/fakethings/kallsyms:/proc/kallsyms"
command+=" -b debian-fs/proc/fakethings/keys:/proc/keys"
command+=" -b debian-fs/proc/fakethings/key-users:/proc/key-users"
command+=" -b debian-fs/proc/fakethings/kpageflags:/proc/kpageflags"
command+=" -b debian-fs/proc/fakethings/loadavg:/proc/loadavg"
command+=" -b debian-fs/proc/fakethings/locks:/proc/locks"
command+=" -b debian-fs/proc/fakethings/misc:/proc/misc"
command+=" -b debian-fs/proc/fakethings/modules:/proc/modules"
command+=" -b debian-fs/proc/fakethings/pagetypeinfo:/proc/pagetypeinfo"
command+=" -b debian-fs/proc/fakethings/partitions:/proc/partitions"
command+=" -b debian-fs/proc/fakethings/sched_debug:/proc/sched_debug"
command+=" -b debian-fs/proc/fakethings/softirqs:/proc/softirqs"
command+=" -b debian-fs/proc/fakethings/timer_list:/proc/timer_list"
command+=" -b debian-fs/proc/fakethings/uptime:/proc/uptime"
command+=" -b debian-fs/proc/fakethings/vmallocinfo:/proc/vmallocinfo"
command+=" -b debian-fs/proc/fakethings/zoneinfo:/proc/zoneinfo"
command+=" -b /sdcard:/media/sd"
command+=" -b /storage/FE3F-10FC:/media/sd1"
## uncomment the following line to have access to the home directory of termux
#command+=" -b /data/data/com.termux/files/home:/root"
## uncomment the following line to mount /sdcard directly to / 
command+=" -b /sdcard"
command+=" -w /root"
command+=" /usr/bin/env "
command+=" HOME=/root"
command+=" PATH=/usr/local/sbin:/usr/local/bin:/bin:/usr/bin:/sbin:/usr/sbin:/usr/games:/usr/local/games"
command+=" TERM=$TERM"
command+=" LANG=zh_CN.UTF-8"
exec $command /sbin/init-proot & ; login $@

help me check this…
zsh doesn't want to start it
and proot quited when it startup
but i still can start my container by other script…

**Reply by michalbednarski** [reply]
I think you want last line to be `exec $command /bin/sh -c "init -i & ; login"`

**Reply by dontknowhy** [reply]
> I think you want last line to be `exec $command /bin/sh -c "init -i & ; login"`

/usr/bin/sh: 1: Syntax error: ";" unexpected
…

**Reply by dontknowhy** [reply]
> > I think you want last line to be `exec $command /bin/sh -c "init -i & ; login"`
> 
> /usr/bin/sh: 1: Syntax error: ";" unexpected
> …

i started the container…
but…
![Screenshot_20220520_185002.jpg](https://user-images.githubusercontent.com/85813508/169512869-123f5f11-f441-4e91-bd10-0e991f30f754.jpg)
my termux IS BROKEN!!!
how to disable this selinux… or how can i disable this output…(tried >>/dev/null)

**Reply by dontknowhy** [reply]
> I think you want last line to be `exec $command /bin/sh -c "init -i & ; login"`

now it's working
and terminal working fine…
but the selinux mount failed text still can't fix it(i need to login twice because login will think this mesage is my password)
i tried to disable selinux by creating a fake config file in /etc/selinux/config
but dpkg will crash because its connection to the information channel is disconnected (translated from Chinese)
how can i clear this mesage?(it will make terminal run into a problem)

**Reply by michalbednarski** [reply]
I guess it'd be
`exec $command /bin/sh -c "init -i < /dev/null > /dev/null 2> /dev/null & ; login"`

Though I think you're going too far and need to step back and answer what you need `init` for in first place

**Reply by dontknowhy** [reply]
> I guess it'd be
> `exec $command /bin/sh -c "init -i < /dev/null > /dev/null 2> /dev/null & ; login"`
> 
> Though I think you're going too far and need to step back and answer what you need `init` for in first place

i only want a working service manager and it's working like a normal computer…
also this command doesn't work:
![Screenshot_20220523_175523.jpg](https://user-images.githubusercontent.com/85813508/169794500-317e9cc9-1b85-45c5-bb41-ca2d6db597d9.jpg)
i tried bash but it doesn't work too…

**Comment by JanuszChmiel** [comment]
Some important remarks related to services.
Some services depend on real Linux kernel so it is impossible to use them from Proot.
Some services can depend on standard Linux.
So before loading some services from Bash shell or from other shell.
It is necessary to contact The specific service author and ask him, if his service depend on some specific Linux kernel system call or kernel module, which is not Thepart of Android kernel.
I personally do not start services in Chroot enviroinment on Android device.
I simply run apps, desktop environments, I can build and compile apps.
I would like to have dbus-broker functioning on Fedora37. Not standard dbus, but really dbus-broker. It is preconfigured professionally when user use The official Fedora Linux installer.
But The question is, if dbus-broker do not depend on some specific Linux kernel system calls.
Some apps such as various modern Linux repositoryes containers refuse to run on non rooted Android devices.
For example Flatpak.
So it is impossible to anticipate, that chrooted Linux container which are being executed by Android non rooted devices by using Proot will work as a Linux distros which ou are boot on virtual machine or directly from some bootable arbitrary media.

From my personal experience.
Proot work The best on devices, where Proot do not have to use too much work arounds to overcome issues with some Android kernels and system configurations.
Some devices have wrongly implemented Pokedata system call, some devices are having troubles with other file system features or system calls.

Some devices have encryption enabled by default, so performance is also slower.
It is complex problematic.
